### PR TITLE
Make DB upgrade queries idempotent for convenience

### PR DESCRIPTION
**Pull request by cl-k-takahashi**
## Description

Make DB upgrade queries idempotent for convenience

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

Manually deleted 4.13.1.0 row from version table and re-started CloudStack management server.

### CODE REVIEW

**Comment by cl-k-takahashi** [comment]
In my environment, `ALTER TABLE `region` DROP INDEX IF EXISTS `id_3`;` success but the travis ci seems to be failing...
```
mysql> select version();
+----------------------------------+
| version()                        |
+----------------------------------+
| 10.1.44-MariaDB-0ubuntu0.18.04.1 |
+----------------------------------+
1 row in set (0.00 sec)
```

**Comment by cl-k-takahashi** [comment]
It seems that MariaDB supports `DROP INDEX IF EXISTS` but MySQL doesn't.

File: engine/schema/src/main/resources/META-INF/db/schema-41310to41400-cleanup.sql [hunk 1]
```diff
@@ -25,4 +25,4 @@ DELETE FROM `cloud`.`configuration` WHERE name = 'host.maintenance.retries';
 UPDATE `cloud`.`storage_pool` SET path="/var/lib/libvirt/images" WHERE path="/var/lib/libvirt/images/";
 
 -- remove (one of) duplicate unique indexes from Region table
-ALTER TABLE `region` DROP INDEX `id_3`;
+ALTER TABLE `region` DROP INDEX IF EXISTS `id_3`;
```
Comment (GabrielBrascher) [inline engine/schema/src/main/resources/META-INF/db/schema-41310to41400-cleanup.sql:28]: MySQL does not support `DROP INDEX IF EXISTS`, at least as far as I know there is no version supporting. If there is, can you please point it here?

Comment (rhtyd) [reply engine/schema/src/main/resources/META-INF/db/schema-41310to41400-cleanup.sql:28]: I ran this, it worked:
```
> ALTER TABLE `region` DROP INDEX IF EXISTS `id_3`;
Query OK, 0 rows affected, 1 warning (0.002 sec)
Records: 0  Duplicates: 0  Warnings: 1

```

File: engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql [hunk 2]
```diff
@@ -49,9 +49,9 @@ CREATE TABLE IF NOT EXISTS `cloud`.`backup_offering` (
   CONSTRAINT `fk_backup_offering__zone_id` FOREIGN KEY (`zone_id`) REFERENCES `data_center` (`id`) ON DELETE CASCADE
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
 
-ALTER TABLE `cloud`.`vm_instance` ADD COLUMN `backup_offering_id` bigint unsigned DEFAULT NULL COMMENT 'ID of backup offering';
-ALTER TABLE `cloud`.`vm_instance` ADD COLUMN `backup_external_id` varchar(255) DEFAULT NULL COMMENT 'ID of external backup job or container if any';
-ALTER TABLE `cloud`.`vm_instance` ADD COLUMN `backup_volumes` text DEFAULT NULL COMMENT 'details of backedup volumes';
+CALL `cloud`.`IDEMPOTENT_ADD_COLUMN`('cloud.vm_instance','backup_offering_id', 'bigint unsigned DEFAULT NULL COMMENT \'ID of backup offering\'');
+CALL `cloud`.`IDEMPOTENT_ADD_COLUMN`('cloud.vm_instance','backup_external_id', 'varchar(255) DEFAULT NULL COMMENT \'ID of external backup job or container if any\'');
+CALL `cloud`.`IDEMPOTENT_ADD_COLUMN`('cloud.vm_instance','backup_volumes', 'text DEFAULT NULL COMMENT \'details of backedup volumes\'');
```
Comment (rhtyd) [inline engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql:54]: Procedure `IDEMPOTENT_ADD_COLUMN` may not be declared in this sql file.

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@GabrielBrascher are you satisfied with @rhtyd 's answer?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM

File: engine/schema/src/main/resources/META-INF/db/schema-41310to41400-cleanup.sql (hunk 1 above)
Comment (GabrielBrascher) [reply engine/schema/src/main/resources/META-INF/db/schema-41310to41400-cleanup.sql:28]: @rhtyd are yiou using mongoDB or mariaDB?

I just execcuted it on a Mysql and it indeed fails:
```
myysql> ALTER TABLE `region` DROP INDEX IF EXISTS `id_3`;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'IF EXISTS `id_3`' at line 1
```

**Review by GabrielBrascher — CHANGES_REQUESTED** [review]

File: engine/schema/src/main/resources/META-INF/db/schema-41310to41400-cleanup.sql (hunk 1 above)
Comment (GabrielBrascher) [reply engine/schema/src/main/resources/META-INF/db/schema-41310to41400-cleanup.sql:28]: Mysql documentation (https://dev.mysql.com/doc/refman/5.7/en/drop-index.html):
```
DROP INDEX [ONLINE | OFFLINE] index_name ON tbl_name
    [algorithm_option | lock_option] ...

algorithm_option:
    ALGORITHM [=] {DEFAULT | INPLACE | COPY}

lock_option:
    LOCK [=] {DEFAULT | NONE | SHARED | EXCLUSIVE}
```

MariaDB:
```
DROP INDEX [IF EXISTS] index_name ON tbl_name 
    [WAIT n |NOWAIT]
    [algorithm_option | lock_option] ...

algorithm_option:
    ALGORITHM [=] {DEFAULT|INPLACE|COPY|NOCOPY|INSTANT}

lock_option:
    LOCK [=] {DEFAULT|NONE|SHARED|EXCLUSIVE}
```

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
Hi @cl-k-takahashi Please check the comments above, and address the changes required (if any). Thanks.

**Comment by nvazquez** [comment]
Ping @cl-k-takahashi can you please check the open comments?

**Comment by nvazquez** [comment]
Ping @cl-k-takahashi can you please check the open comments?

**Comment by rohityadavcloud** [comment]
@cl-k-takahashi  thank you for your PR contribution. Due to inactivity on the PR for more than two years and unresolved conflicts, I'm closing this PR. If you've time and interest please do raise a new PR, thanks again.

**Pull request closed without merging**